[WIP] `hash.tsv` new format plus linting and auto-adding hashes

fabianegli opened the pull request:
The [`hash.tsv`](https://github.com/BioContainers/multi-package-containers/blob/master/combinations/hash.tsv) already contains a list of all the mulled containers, but it does not include their hashes. It would be very convenient if it did, because then the helper tools like the one [here](https://biocontainers.pro/multipackage) would not have to re-calculate them.

This PR is intended to solve the following issues:

- [ ] https://github.com/BioContainers/multi-package-containers/issues/2514
- [ ] https://github.com/BioContainers/multi-package-containers/issues/2521

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

fabianegli committed:
extract function for package sorting

fabianegli committed:
add venv to gitignore

fabianegli committed:
rename function to parse a targets string

nsoranzo commented on lib/galaxy/tool_util/deps/mulled/mulled_build.py:
```diff
@@ -513,7 +513,7 @@ def add_single_image_arguments(parser):
 
 
-def target_str_to_targets(targets_raw):
+def parse_targets(targets_raw):
```
Not sure the renaming is worth the risk of breaking downstream code using the `galaxy-tool-util` package?

fabianegli replied:
Maybe not. This should - if at all - be done in steps with a deprecation warning announcing the change.

So why did I even do it? Its functionality - or rather the pieces that make up its functionality - are not reused as widely as it could be. When you run `git grep -n '.split(' *.py` in the `mulled` folder, you'll see that the parsing is reinvented over and over again. And I am hesitant to add another version of the parsing of container definitions in a new linting module. So it the intention is not just to rename this function, but to decompose it into reusable parts and reuse them whenever a string is parsed that defines a mulled container.

fabianegli committed:
Revert "rename function to parse a targets string"

This reverts commit c9bec38366c2728ab7d5f24b75c33b784e8b9c4e.

fabianegli committed:
do not change test setup in test

fabianegli committed:
improve legibility

- removed unnecessary nesting level from TESTCASES
- put parametrization on multiple lines

fabianegli committed:
extract method  generate_targets_from_file

fabianegli committed:
add function to build string from target

nsoranzo commented on lib/galaxy/tool_util/deps/mulled/util.py:
```diff
@@ -239,6 +239,17 @@ def _simple_image_name(targets, image_build=None):
     return f"{target.package_name}{suffix}"
 
 
+def sort_build_targets(build_targets):
```
```suggestion
def sort_build_targets(build_targets: List[Target]) -> List[Target]:
```

fabianegli committed:
iterate over keys instead of items

fabianegli committed:
add type annotations to sort_build_targets

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>

fabianegli committed:
fix whitespace linting issues

fabianegli committed:
fix formatting

fabianegli committed:
Merge branch 'dev' into hash-tsv-utils
